play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala

```diff
@@ -164,5 +158,130 @@ object StandaloneAhcWSClient {
+          }
+          if (isEmpty(uri.getHost)) {
+            throw new IllegalArgumentException(req.url + " could not be parsed into a proper Uri, missing host")
+          }
+
+          req
+        case Failure(_) =>
+          // URI parsing error. Sometimes recoverable by UriEncoder.FIXING
+          repair(req)
+      }
+    }
+  }
+
+  /**
+   * Encodes the URI to [[Uri]] and runs it through the same [[UriEncoder.FIXING]]
+   * that async-http-client uses before executing it.
+   */
+  @throws[IllegalArgumentException]("if the url is unrepairable")
+  private def repair(req: StandaloneAhcWSRequest): StandaloneWSRequest = {
+    try {
+      val encodedAhcUri: Uri = toUri(req)
+      val javaUri = encodedAhcUri.toJavaNetURI
+      setUri(req, encodedAhcUri.withNewQuery(null).toUrl, Option(javaUri.getRawQuery))
+    } catch {
+      case NonFatal(t) =>
+        throw new IllegalArgumentException(s"Invalid URL ${req.url}", t)
+    }
+  }
+
+  /**
+   * Builds an AHC [[Uri]] with all parts URL encoded by [[UriEncoder.FIXING]].
+   * Combines query params from both [[StandaloneWSRequest.url]] and [[StandaloneWSRequest.queryString]].
+   */
+  private def toUri(req: StandaloneWSRequest): Uri = {
+    val combinedUri: Uri = {
+      val uri = Uri.create(req.url)
+
+      val paramsMap = req.queryString
+      if (paramsMap.nonEmpty) {
+        val query: String = combineQuery(uri.getQuery, paramsMap)
+        uri.withNewQuery(query)
+      } else {
+        uri
+      }
+    }
+
+    // FIXING.encode() encodes ONLY unencoded parts, leaving encoded parts untouched.
+    UriEncoder.FIXING.encode(combinedUri, Collections.emptyList())
+  }
+
+  private def combineQuery(query: String, params: Map[String, Seq[String]]): String = {
+    val sb = new StringBuilder
+    // Reminder: ahc.Uri.query does include '?' (unlike java.net.URI)
+    if (query != null) {
+      sb.append(query)
+    }
+
+    for {
+      (key, values) <- params
+      value <- values
+    } {
+      if (sb.nonEmpty) {
+        sb.append('&')
+      }
+      sb.append(key)
+      if (value.nonEmpty) {
+        sb.append('=').append(value)
+      }
+    }
+
+    sb.toString
+  }
+
+  /**
+   * Replace the [[StandaloneWSRequest.url]] and [[StandaloneWSRequest.queryString]]
+   * with the values of [[uri]], discarding originals.
+   */
+  private def setUri(
+    req: StandaloneAhcWSRequest,
+    urlNoQueryParams: String,
+    encodedQueryString: Option[String]): StandaloneWSRequest = {
+    val queryParams: List[(String, String)] = for {
+      queryString <- encodedQueryString.toList
+      // https://stackoverflow.com/a/13592567 for all of this.
+      pair <- queryString.split('&')
+      idx = pair.indexOf('=')
+      key = URLDecoder.decode(if (idx > 0) pair.substring(0, idx) else pair, "UTF-8")
+      value = if (idx > 0) URLDecoder.decode(pair.substring(idx + 1), "UTF-8") else ""
+    } yield key -> value
+
+    req
+      .withUrl(urlNoQueryParams)
```

Comment:
One concern: `withUrl()` calls `normalize()` which could lead to infinite mutual recursion if I missed a case. I'd prefer to call the `StandaloneAhcWSRequest.copy()` escape hatch here, but it requires access to the implicit `Materializer` which is not accessible.

Some ideas in order of preference (least -> most):
* Leave it as is and let users report cases we missed. Should happen pretty quickly, but ugh.
* Remove the check from `withUrl()`. That'd weaken the integrity of `StandaloneAhcWSRequest`, although it's already comprisable with `copy()` if you provide your own `Materializer.`
* Change visibility to `StandaloneAhcWSRequest(...)(implicit private[ahc] val materializer: Materializer)` and do the `copy()`.

Reply:
I'm going to go ahead and push up a second commit with the last approach. If you like it, I'll squash. Otherwise, I'll drop it.

Reply:
Done. Thoughts?